src/libraries/System.Private.CoreLib/src/System/Threading/RegisteredWaitHandle.Windows.cs

```diff
@@ -0,0 +1,206 @@
+
+            _callbackHelper = callbackHelper;
+            _millisecondsTimeout = millisecondsTimeout;
+            _repeating = repeating;
+
+            // Allocate _gcHandle and _tpWait as the last step and make sure they are never leaked
+            _gcHandle = GCHandle.Alloc(this);
+
+            _tpWait = Interop.Kernel32.CreateThreadpoolWait(&RegisteredWaitCallback, (IntPtr)_gcHandle, IntPtr.Zero);
+
+            if (_tpWait == IntPtr.Zero)
+            {
+                _gcHandle.Free();
+                throw new OutOfMemoryException();
+            }
+        }
+
+        internal RegisteredWaitHandle(WaitHandle waitHandle, _ThreadPoolWaitOrTimerCallback callbackHelper,
+            int millisecondsTimeout, bool repeating)
+        {
+            Debug.Assert(!ThreadPool.UseWindowsThreadPool);
+            GC.SuppressFinalize(this);
+
+            Thread.ThrowIfNoThreadStart();
+            _waitHandle = waitHandle.SafeWaitHandle;
+            _callbackHelper = callbackHelper;
+            _signedMillisecondsTimeout = millisecondsTimeout;
+            _repeating = repeating;
+            if (!IsInfiniteTimeout)
+            {
+                RestartTimeout();
+            }
+        }
+
+        private static AutoResetEvent? s_cachedEvent;
+
+        private static readonly LowLevelLock s_callbackLock = new LowLevelLock();
+
+        /// <summary>
+        /// The callback to execute when the wait on <see cref="Handle"/> either times out or completes.
+        /// </summary>
+        internal _ThreadPoolWaitOrTimerCallback? Callback
+        {
+            get => _callbackHelper;
+        }
+
+        /// <summary>
+        /// The <see cref="SafeWaitHandle"/> that was registered.
+        /// </summary>
+        internal SafeWaitHandle Handle
+        {
+            // get => _waitHandle;
+            get
+            {
+                Debug.Assert(_waitHandle != null);
+                return _waitHandle;
+            }
+        }
+
+        /// <summary>
+        /// The time this handle times out at in ms.
+        /// </summary>
+        internal int TimeoutTimeMs { get; private set; }
+
+        internal int TimeoutDurationMs
+        {
+            get => _signedMillisecondsTimeout;
+        }
+
+        internal bool IsInfiniteTimeout => TimeoutDurationMs == -1;
+
+        /// <summary>
+        /// Whether or not the wait is a repeating wait.
+        /// </summary>
+        internal bool Repeating
+        {
+            get => _repeating;
+        }
+
+        /// <summary>
+        /// The <see cref="WaitHandle"/> the user passed in via <see cref="Unregister(WaitHandle)"/>.
+        /// </summary>
+        private SafeWaitHandle? UserUnregisterWaitHandle { get; set; }
+
+        private IntPtr UserUnregisterWaitHandleValue { get; set; }
+
+        private static IntPtr InvalidHandleValue => new IntPtr(-1);
+
+        internal bool IsBlocking => UserUnregisterWaitHandleValue == InvalidHandleValue;
```

Comment:
A number of these and perhaps other properties are either only used by the portable implementation or by the Windows thread pool implementation. Would be good to move them into the relevant files. There may be some further consolidation that could be done later in a different PR.